test/filters/test_convolution.cpp

```diff
@@ -41,3120 +41,371 @@
+    EXPECT_FLOAT_EQ ((*output) (i, 47).intensity, output_results[i * 2 + 1]);
+  }
+}
+
+TEST (Convolution, convolveRowsRGB)
+{
+  // input
+  Eigen::ArrayXf filter(7);
+  filter << 0.00443305, 0.0540056, 0.242036, 0.39905, 0.242036, 0.0540056, 0.00443305;
+  auto input = pcl::make_shared<PointCloud<RGB>>();
+  input->width = 64;
+  input->height = 48;
+  input->resize(input->width * input->height);
+  for (std::uint32_t r = 0; r < input->height; r++)
+    for (std::uint32_t c = 0; c < input->width; c++)
+    {
+      float x1 = -2.0f + (4.0f / (float)input->width) * (float)c;
+      float y1 = -2.0f + (4.0f / (float)input->height) * (float)r;
+      float x2 = -M_PI + (2.0f * M_PI / (float)input->width) * (float)c;
+      float y2 = -2.0f + (4.0f / (float)input->height) * (float)r;
+      float z = x1 * exp(-(x1 * x1 + y1 * y1)) * 2.5f + sin(x2) * sin(y2);
+      (*input) (c, r) = interpolate_color(-1.6f, 1.6f, z);
+    }
+
+  // filter
+  auto output = pcl::make_shared<PointCloud<RGB>>();
+  pcl::filters::Convolution<RGB, RGB> convolve;
   convolve.setInputCloud (input);
   convolve.setKernel (filter);
   convolve.convolveRows (*output);
 
-  // first output row
-  Eigen::ArrayXf output_0 (64);
-  output_0[0] = 0; output_0[1] = 0; output_0[2] = 0; output_0[3] = 94.1702; output_0[4] = 93.3588; output_0[5] = 92.8969; output_0[6] = 92.8653; output_0[7] = 92.484; output_0[8] = 92.0451; output_0[9] = 91.6455; output_0[10] = 90.9415; output_0[11] = 90.0539; output_0[12] = 89.3048; output_0[13] = 88.7623; output_0[14] = 88.3004; output_0[15] = 88.0044; output_0[16] = 87.7579; output_0[17] = 87.5424; output_0[18] = 87.4441; output_0[19] = 87.0711; output_0[20] = 86.1525; output_0[21] = 84.9014; output_0[22] = 84.1213; output_0[23] = 83.7668; output_0[24] = 83.5469; output_0[25] = 83.5159; output_0[26] = 83.538; output_0[27] = 83.641; output_0[28] = 83.3946; output_0[29] = 82.7889; output_0[30] = 82.3278; output_0[31] = 82.0983; output_0[32] = 82.2605; output_0[33] = 82.376; output_0[34] = 82.1481; output_0[35] = 81.5513; output_0[36] = 81.1708; output_0[37] = 81.3049; output_0[38] = 81.6366; output_0[39] = 81.5826; output_0[40] = 80.9955; output_0[41] = 80.3633; output_0[42] = 80.1213; output_0[43] = 80.3048; output_0[44] = 80.695; output_0[45] = 80.8831; output_0[46] = 80.6995; output_0[47] = 80.3633; output_0[48] = 80.435; output_0[49] = 81.278; output_0[50] = 83.3201; output_0[51] = 86.7523; output_0[52] = 88.1634; output_0[53] = 83.7249; output_0[54] = 77.6666; output_0[55] = 75.8489; output_0[56] = 77.8468; output_0[57] = 79.8108; output_0[58] = 80.4707; output_0[59] = 80.9244; output_0[60] = 81.8518; output_0[61] = 0; output_0[62] = 0; output_0[63] = 0;
+  // output rows (first and last row)
+  // note: this is because of border policy ignore which sets the border of width = filter_size/2 as NaN for xyz and 0 for rgb values
+  std::array<std::uint8_t, 384> output_results {
```

Comment:
An array of RGB would make more sense. Maybe even an array of array of RGB

Reply:
Yes, I have changed the data type. I would keep the array field like this, because this way the two rows can be mapped compactly as columns. I add an inline comment to the description.